src/stacked_borrows.rs

```diff
@@ -388,10 +365,9 @@ impl<'tcx> Stack {
             // Remove everything above the write-compatible items, like a proper stack. This makes sure read-only and unique
             // pointers become invalid on write accesses (ensures F2a, and ensures U2 for write accesses).
             let first_incompatible_idx = self.find_first_write_incompatible(granting_idx);
-            for item in self.borrows.drain(first_incompatible_idx..).rev() {
-                trace!("access: popping item {:?}", item);
+            for item in self.drain_rev(first_incompatible_idx..self.len()) {
```

Comment:
This is the only call site for `drain_rev`, right?

I used `drain` because it seemed elegant, but AFAIK this is equivalent to
```rust
while self.borrows.len() > first_incompatible_idx {
  let item = self.borrows.pop().unwrap();
  // ...
}
```
so maybe that works better for the stack-with-a-cache?
```

Reply:
I think this API is less obtuse and overall it's a win. It's still hampered a bit by the fact that we need to put the cache back into a valid state once we're done with this operation, but perhaps I've named/documented things a bit better now so it's clearer?